relay/utils/discord_monitor.ts

```diff
@@ -0,0 +1,152 @@
+        }
+        if (env.BASE_GOERLI_RPC && env.LC_BASE_GOERLI) {
+            this.contracts['BaseGoerli'] = {RPC: env.BASE_GOERLI_RPC, Address: env.LC_BASE_GOERLI};
+        }
+        if (env.ARBITRUM_GOERLI_RPC && env.LC_ARBITRUM_GOERLI) {
+            this.contracts['ArbitrumGoerli'] = {RPC: env.ARBITRUM_GOERLI_RPC, Address: env.LC_ARBITRUM_GOERLI};
+        }
+        if (env.SEPOLIA_RPC && env.LC_SEPOLIA) {
+            this.contracts['Sepolia'] = {RPC: env.SEPOLIA_RPC, Address: env.LC_SEPOLIA};
+        }
+        if (env.MUMBAI_RPC && env.LC_MUMBAI) {
+            this.contracts['Mumbai'] = {RPC: env.MUMBAI_RPC, Address: env.LC_MUMBAI};
+        }
+        if (env.FANTOM_RPC && env.LC_FANTOM) {
+            this.contracts['Fantom'] = {RPC: env.FANTOM_RPC, Address: env.LC_FANTOM};
+        }
+        if (env.CHIADO_RPC && env.LC_CHIADO) {
+            this.contracts['Chiado'] = {RPC: env.CHIADO_RPC, Address: env.LC_CHIADO};
+        }
+        if (env.GNOSIS_RPC && env.LC_GNOSIS) {
+            this.contracts['Gnosis'] = {RPC: env.GNOSIS_RPC, Address: env.LC_GNOSIS};
+        }
+        if (env.BSC_RPC && env.LC_BSC) {
+            this.contracts['BSC'] = {RPC: env.BSC_RPC, Address: env.LC_BSC};
+        }
+        if (env.AURORA_RPC && env.LC_AURORA) {
+            this.contracts['Aurora'] = {RPC: env.AURORA_RPC, Address: env.LC_AURORA};
+        }
+    
+        // Instantiate SolidityContracts from .env 
+        for (var endpoint in this.contracts) {
+
+            var curLightClient = new ethers.Contract(
+                this.contracts[endpoint].Address,
+                lc_abi_json.abi,
+                new ethers.providers.JsonRpcProvider(this.contracts[endpoint].RPC) // Provider
+            );
+
+            var curSolidityContract = new SolidityContract(
+                curLightClient,
+                this.contracts[endpoint].RPC,
+            );
+            this.contracts[endpoint].SolidityContract= curSolidityContract;
+        }
+    }
+
+    private async getSlotDelay(contract: SolidityContract) {
+        return await this.beaconApi.getCurrentHeadSlot() - await contract.optimisticHeaderSlot();
+    }
+
+    private async respondToMessage() { //TODO: Implement responsive commands
+        this.client.on(Events.MessageCreate, (message) => {
+            if (message.author.bot) return false; 
+
+                // Nice to have, responsive bot
+                console.log(`Message from ${message.author.username}: ${message.content}`);
+                if (message.content === '') console.log('Empty message') //TODO: Bot can't read user messages
+        });
+    }
+
+    private async dispatchMessage(messageToSend) {
+        this.client.channels.cache.get(env.channel_id).send(messageToSend)
+    }
+
+    public async monitor_delay() {
+
+        for (var contract of Object.keys(this.contracts)) {
+            var name = contract;
+            var delay = await this.getSlotDelay(this.contracts[contract].SolidityContract!)
```

Comment:
Could this line hang forever? Is there a timeout somewhere underneath in the libraries making this web3 requests?

Reply:
The BeaconApi uses the native 'fetch' function from Node.js which doesn't implement timeouts. Ethers.js contracts don't seem to implement timeouts as well, so I will wrap the display_monitor() invocation with a Promise.race() with itself and a timeout and retry this statement 10 times before throwing an error.